### DISCUSSION TITLE
How do I get an array of structs from a json object?

### DISCUSSION THREAD

**Discussion by m4l490n** [nlohmann/json#2270] [Q&A] [answered]
Hello

This is the complement for the question [How do I create a json object from an array of structs?](https://github.com/nlohmann/json/discussions/2266) But in this case, what I need to know is how to do the opposite, this is, fill the structs array from a JSON object.

I have the following JSON:

```json
{
    "array" : [
        {
            "a" : 1,
            "c": true,
            "d":"One",
            "e":1.10
        },
        {
            "a" : 2,
            "c": false,
            "d":"Two",
            "e":2.20
        },
        {
            "a" : 3,
            "c": true,
            "d":"Three",
            "e":3.30
        }
    ]
}
```

And I'm trying to fill the `my_struct_t struct_array[3]` from that JSON. I've tried with this:

```cpp
#include <iostream>
#include "json.hpp"

using namespace std;

using json = nlohmann::json;

typedef struct
{
    uint8_t a;
    bool c;
    string d;
    float e;
} my_struct_t;

void to_json(json& j, const my_struct_t& ms)
{
    j["a"] = ms.a;
    j["c"] = ms.c;
    j["d"] = ms.d;
    j["e"] = ms.e;
}

void from_json(const json& j, my_struct_t& ms)
{
    j.at("a").get_to(ms.a);
    j.at("c").get_to(ms.c);
    j.at("d").get_to(ms.d);
    j.at("e").get_to(ms.e);
}

int main()
{
    my_struct_t struct_array[] = {
        {1, true,  "One",   1.10},
        {2, false, "Two",   2.20},
        {3, true,  "Three", 3.30}
    };

    json result = { {"array", struct_array} };
    cout << result << endl;

    auto JSON = result.get<my_struct_t>();

    return 0;
}
```

The output from `cout` is what I expect:

`{"array":[{"a":1,"c":true,"d":"One","e":1.100000023841858},{"a":2,"c":false,"d":"Two","e":2.200000047683716},{"a":3,"c":true,"d":"Three","e":3.299999952316284}]}`

But I'm getting an exception on `auto JSON = result.get<my_struct_t>();`. The documentation explains how to do this with a single struct, but it is not clear how to do it with an array of structs.

**Comment by nlohmann** [comment, marked as answer]
First, the library will not be able to create a C-style array for you. You will only be able to retrieve a `std::vector`, though. For this, you need to "unpack" the JSON array from the JSON object like this:

```cpp
auto JSON = result["array"].get<std::vector<my_struct_t>>();
```

**Reply by m4l490n** [reply]
Thanks for answering.

I would still have to go from vector to c-style string, so I decided to do this instead:

```cpp
    uint8_t i = 0;
    my_struct_t struct_array_res[3] = {0};

    for(my_struct_t x : result["array"])
    {
        struct_array_res[i++] = x;
    }
```

**Reply by nlohmann** [reply]
Maybe try `for (my_struct& x : result["array"])` to avoid a copy. You may also do something like explained in https://stackoverflow.com/a/2923290.
